lisp/ob-ein.el

```diff
@@ -275,66 +198,87 @@ jupyter kernels.
-;; the context of the session environment.
-(defun org-babel-prep-session:ein (_session _params)
-  "Prepare SESSION according to the header arguments specified in PARAMS."
-  )
-
-(defun org-babel-ein-var-to-template (var)
-  "Convert an elisp var into a string of template source code
- specifying a var of the same value."
-  (format "%S" var))
-
-(defun org-babel-ein-table-or-string (_results)
-  "If the results look like a table, then convert them into an
- Emacs-lisp table, otherwise return the results as a string."
-  )
-
-(defun ein:org-babel-clean-url (url-or-port)
-  (if (search ":" url-or-port)
-      url-or-port
-    (string-to-number url-or-port)))
-
-(defun ein:org-babel-parse-session (session)
-  (if (numberp session)
-      (values (ein:url (format "http://localhost:%s" session)) nil)
-    (let ((session-uri (url-generic-parse-url session)))
-      (cond ((url-fullness session-uri)
-             (values (ein:url (format "%s://%s:%s" (url-type session-uri) (url-host session-uri) (url-port session-uri)))
-                     (url-filename session-uri)))
-            (t (let* ((url-or-port (ein:org-babel-clean-url (car (split-string session "/"))))
-                      (path (ein:join-str "/" (rest (split-string session "/")))))
-                 (values (ein:url (format "http://localhost:%s" url-or-port)) path)))))))
-
-(defcustom ein:org-babel-default-session-name "ein_babel_session.ipynb"
-  "Default name for org babel sessions running ein environments.
-This is the name of the notebook used when no notebook path is
-given in the session parameter."
-  :type '(string :tag "Format string")
-  :group 'ein)
-
-
-(defun org-babel-ein-initiate-session (&optional session kernelspec)
-  "If there is not a current inferior-process-buffer in SESSION then create.
- Return the initialized session."
-  (when (and (stringp session) (string= session "none"))
-    (error "You must specify a notebook or kernelspec as the session variable for ein code blocks."))
-  (multiple-value-bind (url-or-port path) (ein:org-babel-parse-session session)
-    (when (null kernelspec)
-      ;; Now is not the time to be getting kernelspecs.
-      ;; If I must do so, need to inject a deferred callback chain like
-      ;; in ein:notebooklist
-      ;; (if (null (gethash url-or-port ein:available-kernelspecs))
-      ;;     (ein:query-kernelspecs url-or-port))
-      (setq kernelspec (ein:get-kernelspec url-or-port "default")))
-    (cond ((null path)
-           (let* ((name ein:org-babel-default-session-name)
-                  (new-session (format "%s/%s" url-or-port name)))
-             (ein:notebooklist-new-notebook-with-name name kernelspec url-or-port)
-             (org-babel-ein-initiate-session new-session kernelspec)))
-          (t (let ((nb (ein:org-find-or-open-session session kernelspec)))
-               (ein:$notebook-kernel nb))))))
+(defun org-babel-edit-prep:ein (babel-info)
+  (let ((buffer (current-buffer)))
+    (ob-ein--initiate-session
```

Comment:
The signature on this has changed to 3 arguments: session, kernelspec, callback.

Reply:
Thanks, that was a real problem.  I added a test for `C-c '` to exercise edit-prep.